File: ai/chatbot/services/trip_service.py
import sys
from bson import ObjectId
from services.db import mongo


class TripService:

    def get_latest_trip(self, user_id: str):
        print(f">>> get_latest_trip called with user_id: {user_id}", flush=True)
        if mongo.trips is None:
            print("trips collection is None!", flush=True)
            return {}
        try:
            total = mongo.trips.count_documents({})
            print(f"Total trips in DB: {total}", flush=True)

            trip = mongo.trips.find_one(
                {"user": ObjectId(user_id)},
                sort=[("createdAt", -1)]
            )
            print(f"Trip found: {trip is not None}", flush=True)

            if not trip:
                return {}

            trip["_id"] = str(trip["_id"])
            trip["user"] = str(trip["user"])
            return trip

        except Exception as e:
            print(f"Error getting trip: {e}", flush=True)
            return {}

    def get_recent_trips(self, user_id: str, limit: int = 5):
        print(f">>> get_recent_trips called with user_id: {user_id}", flush=True)
        if mongo.trips is None:
            print("trips collection is None!", flush=True)
            return []
        try:
            total = mongo.trips.count_documents({})
            print(f"Total trips in DB: {total}", flush=True)

            trips = list(
                mongo.trips.find(
                    {"user": ObjectId(user_id)},
                    sort=[("createdAt", -1)],
                    limit=limit
                )
            )
            print(f"Trips found: {len(trips)}", flush=True)

            for t in trips:
                t["_id"] = str(t["_id"])
                t["user"] = str(t["user"])

            return trips

        except Exception as e:
            print(f"Error getting trips: {e}", flush=True)
            return []

File: backend/src/modules/trip/trip.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { TripRepository, VehicleRepository } from 'src/DB';
import { UpdateTripDTO } from './dto/trip.dto';
import { Types } from 'mongoose';
import { ITrip } from 'src/common';
import { AiService } from './ai/ai.service';

@Injectable()
export class TripService {
  constructor(
    private readonly tripRepository: TripRepository,
    private readonly vehicleRepository: VehicleRepository,
    private readonly aiService: AiService,
  ) {}

  formatForChatbot(trip: any): string {
    return `
🚗 Trip Summary:
Distance: ${trip?.trip_summary?.distance_km ?? 0} km
Avg Speed: ${trip?.trip_summary?.avg_speed ?? 0} km/h

🧠 Driver Score: ${trip?.driving_behavior?.driver_score ?? 'N/A'}

🔧 Vehicle Health: ${trip?.vehicle_health?.health_status ?? 'Unknown'}

⛽ Fuel: ${trip?.fuel_efficiency?.efficiency_label ?? 'Unknown'}
`;
  }

  async createTrip(rawData: any, userId: string) {
    const vehicle = await this.vehicleRepository.findOne({
      filter: { userId: new Types.ObjectId(userId) },
    });

    if (!vehicle) {
      throw new NotFoundException('Vehicle not found');
    }

    const computedTrip = await this.aiService.buildTrip(rawData, vehicle);

    const tripToSave = {
      user: new Types.ObjectId(userId),
      confirmed: true,
      trip_id: rawData.trip_id,
      date: rawData.date,

      ...computedTrip,
    };

    return this.tripRepository.create({
      data: [tripToSave],
    });
  }

  async getTrip(tripId: string) {
    const trip = await this.tripRepository.findOne({
      filter: { _id: new Types.ObjectId(tripId) },
      options: {
        populate: [{ path: 'user', select: 'email username' }],
      },
    });

    if (!trip) {
      throw new NotFoundException('trip not found');
    }

    return trip;
  }

  async getUserTrips(userId: string, page: number = 1, size: number = 5) {
    return this.tripRepository.paginate({
      filter: { user: new Types.ObjectId(userId) },
      options: { sort: { createdAt: -1 } },
      page,
      size,
    });
  }

  async updateTrip(tripId: string, data: UpdateTripDTO): Promise<string> {
    const trip = await this.tripRepository.findOne({
      filter: { _id: new Types.ObjectId(tripId) },
    });

    if (!trip) {
      throw new NotFoundException('trip not found');
    }

    for (const key of Object.keys(data)) {
      if (data[key] !== undefined) {
        trip[key] = data[key];
      }
    }

    await trip.save();

    return 'Done';
  }

  async confirmTrip(tripId: string): Promise<string> {
    const trip = await this.tripRepository.findOne({
      filter: { _id: new Types.ObjectId(tripId) },
    });

    if (!trip) {
      throw new NotFoundException('trip not found');
    }

    trip.confirmed = true;
    await trip.save();

    return 'Done';
  }

  async deleteTrip(tripId: string): Promise<string> {
    const deleted = await this.tripRepository.deleteOne({
      filter: { _id: new Types.ObjectId(tripId) },
    });

    if (!deleted.deletedCount) {
      throw new NotFoundException('trip not found');
    }

    return 'Done';
  }

  async getLatestTrip(userId: string) {
    const [trip] = await this.tripRepository.find({
      filter: { user: new Types.ObjectId(userId) },
      options: {
        sort: { createdAt: -1 },
        limit: 1,
      },
    });

    if (!trip) throw new NotFoundException('No trips');

    return trip;
  }

  async getLatestTripFormatted(userId: string) {
    const trip = await this.getLatestTrip(userId);
    return this.formatForChatbot(trip);
  }

  async getWeeklyReport(userId: string) {
    const trips: ITrip[] = await this.tripRepository.find({
      filter: { user: new Types.ObjectId(userId) },
    });

    if (!trips.length) {
      return {
        total_trips: 0,
        total_distance: 0,
        avg_driver_score: 0,
        avg_fuel: 0,
      };
    }

    const total_trips = trips.length;

    const total_distance = trips.reduce(
      (sum, t) => sum + (t.trip_summary?.distance_km || 0),
      0,
    );

    const avg_driver_score =
      trips.reduce(
        (sum, t) => sum + (t.driving_behavior?.driver_score || 0),
        0,
      ) / total_trips;

    const avg_fuel =
      trips.reduce(
        (sum, t) => sum + (t.fuel_efficiency?.actual_fuel_l_100km || 0),
        0,
      ) / total_trips;

    return {
      total_trips,
      total_distance,
      avg_driver_score,
      avg_fuel,
    };
  }
}
